File: python/graph_caster/contract/document.py
# Copyright GraphCaster. All Rights Reserved.
# Authoritative source: schemas/graph-document.schema.json (v1.20)
#
# This module mirrors the JSON Schema shape as pydantic models. It is a
# hand-written mirror until scripts/codegen.sh (planned) wires
# datamodel-code-generator. The schema remains the single source of truth:
# any drift between this file and the schema is a bug.
#
# Forward-compatibility: every model sets `extra="allow"` so unknown fields
# round-trip unchanged. This matches `additionalProperties: true` in the
# schema and the existing UI behaviour.
#
# Public API:
#   Document  - top-level graph document
#   Node      - graph node (id, type, position, data, mode?, parentId?)
#   Edge      - graph edge with both camelCase (canonical) and snake_case
#               legacy handle aliases preserved verbatim
#   Viewport  - canvas viewport { x, y, zoom }
#   Meta      - document meta block (schemaVersion, graphId, title, author)

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class _Contract(BaseModel):
    """Common base: forward-compat (`extra=allow`) + accept aliases."""

    model_config = ConfigDict(
        extra="allow",
        populate_by_name=True,
    )


class Viewport(_Contract):
    x: float | None = None
    y: float | None = None
    zoom: float | None = None


class Meta(_Contract):
    schemaVersion: int | None = None
    graphId: str | None = None
    title: str | None = None
    author: str | None = None


class Node(_Contract):
    id: str
    type: str
    position: dict[str, Any] = Field(default_factory=lambda: {"x": 0.0, "y": 0.0})
    data: dict[str, Any] = Field(default_factory=dict)
    # Execution mode. Mirrors models.NODE_MODES: normal|bypass|mute|disabled.
    # Optional in schema, defaults applied at the legacy from_dict layer.
    mode: str | None = None
    # Optional parent node id for group/frame membership.
    parentId: str | None = None


class Edge(_Contract):
    id: str
    source: str
    target: str
    # camelCase = canonical (schema); snake_case retained as forward-compat
    # alias since old graphs/ on disk may still use these keys until a
    # migration scrubs them. parseDocument.ts has the matching JS fallback.
    sourceHandle: str | None = None
    targetHandle: str | None = None
    source_handle: str | None = None
    target_handle: str | None = None
    condition: str | None = None
    data: dict[str, Any] | None = None


class Document(_Contract):
    schemaVersion: int | None = None
    graphId: str | None = None
    meta: Meta | None = None
    viewport: Viewport | None = None
    inputs: Any | None = None
    outputs: Any | None = None
    nodes: list[Node] = Field(default_factory=list)
    edges: list[Edge] = Field(default_factory=list)

File: ui/src/contract/document.ts
// Copyright GraphCaster. All Rights Reserved.
// Authoritative source: schemas/graph-document.schema.json (v1.20)
//
// Hand-written mirror of the JSON Schema until scripts/codegen.sh (planned)
// runs json-schema-to-typescript. The schema is the single source of truth;
// drift between this file and the schema is a bug.
//
// Forward-compatibility: extra unknown fields are preserved via
// `[key: string]: unknown` index signatures, mirroring
// `additionalProperties: true` in the schema.
//
// These types are intentionally identical in shape to the legacy
// ui/src/graph/types.ts so consumers can migrate one import at a time.

export type ContractViewport = {
  x?: number;
  y?: number;
  zoom?: number;
  [key: string]: unknown;
};

export type ContractMeta = {
  schemaVersion?: number;
  graphId?: string;
  title?: string;
  author?: string;
  [key: string]: unknown;
};

export type ContractNode = {
  id: string;
  type: string;
  position?: { x?: number; y?: number; [key: string]: unknown };
  /** Optional parent frame (comment/group) id. */
  parentId?: string;
  data?: Record<string, unknown>;
  /**
   * Execution mode. Mirrors python contract.document.Node.mode:
   * "normal" | "bypass" | "mute" | "disabled" (or absent).
   */
  mode?: string;
};

export type ContractEdgeData = {
  routeDescription?: string;
  /** F18: optional override of effective port kind on source side. */
  sourcePortKind?: string;
  /** F18: optional override on target side. */
  targetPortKind?: string;
  [key: string]: unknown;
};

export type ContractEdge = {
  id: string;
  source: string;
  target: string;
  // camelCase is canonical. snake_case kept as forward-compat alias
  // until graphs/ are migrated to camelCase.
  sourceHandle?: string | null;
  targetHandle?: string | null;
  source_handle?: string | null;
  target_handle?: string | null;
  condition?: string | null;
  data?: ContractEdgeData;
};

export type ContractDocument = {
  schemaVersion?: number;
  /** Legacy: mirrors meta.graphId. */
  graphId?: string;
  meta?: ContractMeta;
  inputs?: unknown;
  outputs?: unknown;
  viewport?: ContractViewport;
  nodes?: ContractNode[];
  edges?: ContractEdge[];
  [key: string]: unknown;
};
